File: src/converter/plugins/ImportInfoPlugin.ts
import {ConverterPlugin} from "../plugin.js";
import ts, {ModuleDeclaration, Node, Program} from "typescript";
import {ConverterContext} from "../context.js";
import {Render} from "../render.js";
import {GeneratedFile} from "../generated.js";
import {ImportInfo} from "../../structure/import/collectImportInfo.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";

export const importInfoServiceKey = Symbol()

export class ImportInfoService {
    constructor(
        private readonly program: Program,
        private readonly importInfo: ImportInfo,
    ) {
    }

    resolveImports(sourceFileName: string, node: ModuleDeclaration | undefined): string[] {
        if (node) {
            return this.importInfo.get(node) ?? []
        } else {
            const sourcefile = this.program.getSourceFile(sourceFileName)
            if (!sourcefile) return []

            return this.importInfo.get(sourcefile) ?? []
        }
    }
}

export class ImportInfoPlugin implements ConverterPlugin {
    private readonly importInfoService: ImportInfoService;

    constructor(
        program: Program,
        importInfo: ImportInfo
    ) {
        this.importInfoService = new ImportInfoService(program, importInfo);
    }

    generate(): GeneratedFile[] {
        return [];
    }

    render(node: Node, context: ConverterContext, next: Render): string | null {
        if (ts.isImportDeclaration(node)) {
            const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
            checkCoverageService?.deepCover(node)

            return ""
        }

        return null;
    }

    traverse(node: Node): void {
    }

    setup(context: ConverterContext): void {
        context.registerService(importInfoServiceKey, this.importInfoService)
    }
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/ImportInfoPlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.ConverterContext
import io.github.sgrishchenko.karakum.extension.ConverterPlugin
import io.github.sgrishchenko.karakum.extension.GeneratedFile
import io.github.sgrishchenko.karakum.extension.Render
import io.github.sgrishchenko.karakum.structure.import.ImportInfo
import js.array.ReadonlyArray
import js.symbol.Symbol
import typescript.ModuleDeclaration
import typescript.Node
import typescript.Program
import typescript.isImportDeclaration

@OptIn(ExperimentalJsExport::class)
@JsExport
val importInfoServiceKey = Symbol()

class ImportInfoService(
    private val program: Program,
    private val importInfo: ImportInfo,
) {
    fun resolveImports(sourceFileName: String, node: ModuleDeclaration?): ReadonlyArray<String> {
        if (node != null) {
            return importInfo[node] ?: emptyArray()
        } else {
            val sourcefile = program.getSourceFile(sourceFileName) ?: return emptyArray()

            return importInfo[sourcefile] ?: emptyArray()
        }
    }
}

class ImportInfoPlugin(program: Program, importInfo: ImportInfo) : ConverterPlugin<Node> {
    private val importInfoService = ImportInfoService(program, importInfo)

    override fun setup(context: ConverterContext) {
        context.registerService(importInfoServiceKey, importInfoService)
    }

    override fun traverse(node: Node, context: ConverterContext) = Unit

    override fun render(node: Node, context: ConverterContext, next: Render<Node>): String? {
        if (isImportDeclaration(node)) {
            val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
            checkCoverageService?.deepCover(node)

            return ""
        }

        return null
    }

    override fun generate(context: ConverterContext, render: Render<Node>) = emptyArray<GeneratedFile>()
}
